### ISSUE TITLE
Document how to use other span stores in sleuth-zipkin-stream

### ISSUE THREAD

**Issue by adriancole** [spring-cloud/spring-cloud-sleuth#255]
The stream server is a zipkin server, so supports other span stores such as elasticsearch or cassandra. We should document that, probably after modularization work completes (to avoid thrash).

Ex here are the properties, but they will probably move to another library https://github.com/openzipkin/zipkin/blob/master/zipkin-autoconfigure/storage-elasticsearch/src/main/java/zipkin/autoconfigure/storage/elasticsearch/ZipkinElasticsearchStorageProperties.java

https://github.com/openzipkin/zipkin-java/issues/108

**Comment by adriancole** [comment]
ps my opinion on this is to only declare support for a fixed amount of configurations, and those that have been explicitly requested. Ex. document a backend for stream, like kafka, after someone has tested it with sleuth, not before.

For example, we could start by documenting mysql, which we regularly use and actually test with sleuth. Other things could be mentioned as "use at your own risk".

Reason being is that contributors of this project aren't likely to have tried all combinations of tools and lack of testing could lead to a mutually poor support experience.

**Comment by sunnyfun888** [comment]
People who want to use elasticsearch storage can try this:
Use the pom.xml like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
		<groupId>org.springframework.cloud.demo</groupId>
	<artifactId>demo-zipkin-server</artifactId>
	<packaging>jar</packaging>

	<description>Spring Boot Zipkin Server</description>

	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>1.3.7.RELEASE</version>
	</parent>

	<properties>
		<docker.image.prefix>springio</docker.image.prefix>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<java.version>1.8</java.version>
		<sonar.skip>true</sonar.skip>
		<zipkin.version>1.6.0</zipkin.version>
		<elasticsearch.version>2.3.4</elasticsearch.version>
	</properties>
	
	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-sleuth-dependencies</artifactId>
				<version>${project.version}</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
			
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-sleuth-dependencies</artifactId>
				<version>1.0.6.RELEASE</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
			
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-sleuth-sample-test-core</artifactId>
				<version>${project.version}</version>
			</dependency>
			<dependency>
				<groupId>io.zipkin.java</groupId>
				<artifactId>zipkin-autoconfigure-ui</artifactId>
				<version>${zipkin.version}</version>
			</dependency>
			<dependency>
				<groupId>io.zipkin.java</groupId>
				<artifactId>zipkin-autoconfigure-storage-mysql</artifactId>
				<version>${zipkin.version}</version>
			</dependency>
			<dependency>
				<groupId>io.zipkin.java</groupId>
				<artifactId>zipkin</artifactId>
				<version>${zipkin.version}</version>
			</dependency>
			<dependency>
				<groupId>io.zipkin.java</groupId>
				<artifactId>zipkin-server</artifactId>
				<version>${zipkin.version}</version>
			</dependency>
			
		    <dependency>
		    	<groupId>io.zipkin.java</groupId>
		    	<artifactId>zipkin-autoconfigure-storage-elasticsearch</artifactId>
		        <version>${zipkin.version}</version>
		    </dependency>
			
			<dependency>
			  <groupId>org.elasticsearch</groupId>
			  <artifactId>elasticsearch</artifactId>
			  <version>${elasticsearch.version}</version>
			</dependency>
			
		</dependencies>
	</dependencyManagement>

	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-configuration-processor</artifactId>
			<optional>true</optional>
		</dependency>

		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-sleuth-zipkin-stream</artifactId>
		</dependency>
		<dependency>
			<groupId>io.zipkin.java</groupId>
			<artifactId>zipkin-autoconfigure-ui</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.integration</groupId>
			<artifactId>spring-integration-jmx</artifactId>
		</dependency>
		
		<!-- EXAMPLE FOR RABBIT BINDING -->
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-stream-binder-rabbit</artifactId>
		</dependency>

    	<!-- Elasticsearch -->
	    <dependency>
	    	<groupId>io.zipkin.java</groupId>
	    	<artifactId>zipkin-autoconfigure-storage-elasticsearch</artifactId>
	    </dependency>

	</dependencies>

	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
                <configuration>
                    <finalName>demo-zipkin-server</finalName>
                </configuration>
			</plugin>
		</plugins>
	</build>

</project>
```

And the application.yml file:

```YAML
server:
  port: 9411

spring:
  rabbitmq:
    host: ${RABBIT_HOST:localhost}
  sleuth:
    enabled: false
zipkin:
  storage:
    type: elasticsearch
    elasticsearch:
      cluster: ${ES_CLUSTER:elasticsearch}
      hosts: ${ES_HOSTS:localhost:9300}
      index: ${ES_INDEX:zipkin}
      index-shards: ${ES_INDEX_SHARDS:5}
      index-replicas: ${ES_INDEX_REPLICAS:1}
```

Finally the ZipkinServerApplication.java file:

```java
@EnableZipkinStreamServer
@SpringBootApplication
public class ZipkinServerApplication {

	public static void main(String[] args){
		SpringApplication.run(ZipkinServerApplication.class,args);
	}
}
```
Then have fun :-)

What I can say is that it's like a magic cube, you just change the configuration files, then get what you want, it's amazing.

**Comment by marcingrzejszczak** [comment]
That looks fantastic!! Thanks a lot !

**Comment by marcingrzejszczak** [comment]
But this pom has duplications... 

```
<dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-sleuth-dependencies</artifactId>
                <version>${project.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>

            <dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-sleuth-dependencies</artifactId>
                <version>1.0.6.RELEASE</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>

```

```
<dependency>
                <groupId>io.zipkin.java</groupId>
                <artifactId>zipkin-autoconfigure-storage-mysql</artifactId>
                <version>${zipkin.version}</version>
            </dependency>

            <dependency>
                <groupId>io.zipkin.java</groupId>
                <artifactId>zipkin-autoconfigure-storage-elasticsearch</artifactId>
                <version>${zipkin.version}</version>
            </dependency>
```

are you sure all is needed? @sunnyfun888

**Comment by sunnyfun888** [comment]
Try then know ;-)

**Comment by marcingrzejszczak** [comment]
So maybe let me rephrase it... I'm telling you that this

```
<dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-sleuth-dependencies</artifactId>
                <version>${project.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>

            <dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-sleuth-dependencies</artifactId>
                <version>1.0.6.RELEASE</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
```

is wrong. And this

```
<dependency>
                <groupId>io.zipkin.java</groupId>
                <artifactId>zipkin-autoconfigure-storage-mysql</artifactId>
                <version>${zipkin.version}</version>
            </dependency>
```

looks unnecessary

**Comment by sunnyfun888** [comment]
zipkin-autoconfigure-storage-mysql is just for switching to mysql storage, can remove it.

**Comment by sunnyfun888** [comment]
This part can also be removed.
````xml
            <dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-sleuth-dependencies</artifactId>
                <version>${project.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
````

**Comment by burcakulug** [comment]
This is my solution for mysql: http://stackoverflow.com/a/39174580/2733462

**Comment by turu** [comment]
I am considering using spring-cloud-sleuth to setup request tracing in a very large distributed system and I'm wondering if there is a confirmed, working way of using Cassandra as the Span Store? Do you guys know of anyone who managed to do this, did you?

**Comment by marcingrzejszczak** [comment]
ping @adriancole ^^

**Comment by marcingrzejszczak** [comment]
pinging again ;) @adriancole

**Comment by adriancole** [comment]
I suspect this issue is behind this blog, as usually folks will just use the standard zipkin build (unless they are using spring-cloud-stream) https://dzone.com/articles/persisting-zipkin-spanstore-mysql-amp-cassandra

**Comment by ashwgupt** [comment]
Is there is a limitation to use only mysql as the SpanStore? Can we use Oracle to do so?

**Comment by marcingrzejszczak** [comment]
I guess you'd have to build sth like https://github.com/openzipkin/zipkin/tree/master/zipkin-storage/mysql but for Oracle. But @adriancole for sure knows the answer better than me.

**Comment by adriancole** [comment]
There are mysql dialects in use (upsert query). It is possible oracle wont
work and unlikely the open source project would be able to test if it did
(due to licensing). As such formal support for oracle in OpenZipkin would
be very unlikely. If a vendor wanted to make an oracle variant of the
storage library, they could.. however i havent heard anyone else ask for
oracle so that is also unlikely at the moment

On 9 Mar 2017 12:51, "Marcin Grzejszczak" <notifications@github.com> wrote:

I guess you'd have to build sth like https://github.com/openzipkin/
zipkin/tree/master/zipkin-storage/mysql but for Oracle. But @adriancole
<https://github.com/adriancole> for sure knows the answer better than me.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/spring-cloud/spring-cloud-sleuth/issues/255#issuecomment-285319585>,
or mute the thread

**Comment by ashwgupt** [comment]
So in summary, at present only MySql and Cassandra are the 2 proven persistent SpanStores for ZipKin. Is that correct?

**Comment by adriancole** [comment]
elasticsearch is the other

**Comment by ashwgupt** [comment]
Okay, thanks!

**Comment by adriancole** [comment]
no problem!
